client/src/models/itemstreemodel.cpp

```diff
@@ -142,6 +171,80 @@ ItemsTreeModel::ColumnTypes ItemsTreeModel::columnTypes() const
     return d->mColumns;
 }
 
+void ItemsTreeModel::updateBackgrounds()
+{
+    if (rowCount() == 0)
+        return;
+
+    updateBackgrounds(0, rowCount() - 1);
+}
+
+void ItemsTreeModel::updateBackgrounds(int first, int last)
+{
+    Q_ASSERT(first >= 0 && last >= 0 && last < rowCount());
+
+    if (d->mCurrentlyUpdatingBackgrounds)
+        return;
+
+    QScopedValueRollback<bool> reentrancyBlock(d->mCurrentlyUpdatingBackgrounds, true);
+
+    const int lastModifiedDateColumn = columnTypes().indexOf(LastModifiedDate);
+    const int nextStepDateColumn = columnTypes().indexOf(NextStepDate);
+    if (lastModifiedDateColumn == -1 && nextStepDateColumn == -1)
+        return; // no backgrounds to show, nothing to do for this model
+
+    QElapsedTimer timer;
+    timer.start();
+
+    const KColorScheme colors(QPalette::Active);
+    // TODO: Take alternating row colors into account? How?
+    const QColor background = colors.background().color();
+
+    const auto currentDateTime = QDateTime::currentDateTime();
+
+    for (int row = first; row <= last; ++row) {
+        const Item item = index(row, 0).data(ItemRole).value<Item>();
+        Q_ASSERT(item.isValid());
+
+        if (lastModifiedDateColumn != -1) {
+            // LastModifiedDate => Indicate whether opps are still 'warm', cf. FATCRM-109
+            const qint64 secsAfterOppIsCold = 90*24*3600;
+            const auto dateTime = entityData(item, lastModifiedDateColumn, Qt::EditRole).toDateTime();
+            const qint64 secsSinceLastModificationDate = ::clamp(dateTime.secsTo(currentDateTime), 0ll, secsAfterOppIsCold);
+            QColor columnBackground;
+            if (secsSinceLastModificationDate != secsAfterOppIsCold) {
+                const qreal amount = (float)qAbs(secsSinceLastModificationDate - secsAfterOppIsCold) / secsAfterOppIsCold;
+                columnBackground = KColorUtils::mix(background, colors.background(KColorScheme::PositiveBackground).color(), amount);
+            }
+            d->mIdToBackgroundMap[item.id()].lastModifiedDateBackground = columnBackground;
```

Comment:
Why insert invalid colors into the map, when the if() failed? I think it would be better to move this into the if(), to avoid doing that. At lookup time the result is the same, but this prevents using memory for all those items with no background color.

Reply:
It's not as easy. This way we may up ending up with stale entries in said cache. 

Not as noticeable with the lastModifiedDate, but more with the nextStepDate: Consider someone just unsetting the next step date from a next step date one day past the due date: In that case this function would not remove the background color for this item.

I've come up with a more memory conservative solution now :)